"name"-field for human-readable container & container group names in ACI

vholmer opened the pull request:
<!-- Thanks for contributing 🎉! Please ensure the title neatly summarizes the proposed changes. -->

<!-- Overview -->

Naming the containers to something human-readable is great! Which is the idea behind this PR. At my place of work we've long been using this fix to go from a random string of numbers to something a human can read.

### Screenshots
![image](https://user-images.githubusercontent.com/5623158/234023850-5d9ac097-d92b-4e3a-bfe8-4d91070951ae.png)

### Checklist
- [x] References any related issue by including "Closes #<Issue Number>" or "Closes <Issue URL>".
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect-azure/issues/new/choose) first.
- [ ] Includes tests or only affects documentation.
- [ ] Passes `pre-commit` checks.
  - Run `pre-commit install && pre-commit run --all` locally for formatting and linting.
- [x] Summarizes PR's changes in [CHANGELOG.md](https://github.com/PrefectHQ/prefect-azure/blob/main/CHANGELOG.md)

vholmer commented:
Another feature (which I forgot to add at PR creation time, but which is visible in the attached screenshot) that we use is a suffix of "-" + 6 random digits. This is great to have when, for example, multiple flows with the same name are running in different environments.

desertaxle commented on prefect_azure/container_instance.py:
```diff
@@ -214,6 +216,12 @@ class AzureContainerInstanceJob(Infrastructure):
             "defaults to a Prefect base image matching your local versions."
         ),
     )
+    name: Optional[str] = Field(
+        default=None,
+        description=(
+            "Name of the Azure Container Instance."
+        ),
+    )
```
The `Infrastructure` class that `AzureContainerInstanceJob` subclasses already has a `name` attribute, so adding `name` on this classes shouldn't be necessary.
```suggestion
```

desertaxle commented on prefect_azure/container_instance.py:
```diff
@@ -470,7 +478,8 @@ def _configure_container(self) -> Container:
 
         # all container names in a resource group must be unique
-        container_name = str(uuid.uuid4())
+        number_suffix = ''.join(random.choices(string.digits, k=6))
+        container_name = self.name + '-' + number_suffix if self.name else str(uuid.uuid4())
```
Azure's container instances service has some constraints on container names that we should respect. Also, to have a higher assurance of randomness, let's increase the suffix length and add letters.
```suggestion
        slugified_name = slugify(self.name, max_length=52, regex_pattern=r"[^a-zA-Z0-9-]+")
        random_suffix = ''.join(random.choices(string.ascii_lowercase, string.digits, k=10))
        container_name = self.name + '-' + number_suffix
```

`slugify` is a dependency of `prefect` so you should be able to import `slugify` with:
```
import slugify from slugify
```

Also, the `name` attribute for an infrastructure block is populated to the flow run name by default, so it should be populated in all cases.

desertaxle reviewed (changes requested):
Thanks for the contribution @vholmer! To add an entry to the change log you can add a short description of this PR to `CHANGELOG.md` under the Unreleased -> Added section. 

In addition to my other comments, could you also add some tests covering this new functionality?

vholmer commented:
> @desertaxle I've updated the PR as suggested, the regex I added follows the naming restrictions from [here](https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules#microsoftcontainerinstance) minus 9 characters (max 52) to allow for '-' + the 10 length suffix.
> 
> I'm not sure where / how to add the tests for this though! Any suggestions?

I realized slugification already takes care of the annoying bits that my regex attempted to do...reverted back to what you suggested

desertaxle commented:
Looks like the static analysis checks and unit tests are failing.

For the static analysis checks you can a run `pre-commit install && pre-commit run --all` and commit the changes to resolve the failures.

For the unit tests, it looks like `slugify` is failing when `name` is `None`. Since the `AzureContainerInstanceJob` block can be used without a name (it will be populated when used with a deployment) we should check if `name` is `None` and use a UUID if `name` is `None`.

desertaxle reviewed (commented):
Looks like the static analysis checks and unit tests are failing.

For the static analysis checks you can a run pre-commit install && pre-commit run --all and commit the changes to resolve the failures.

For the unit tests, it looks like slugify is failing when name is None. Since the AzureContainerInstanceJob block can be used without a name (it will be populated when used with a deployment) we should check if name is None and use a UUID if name is None.
